Frontend/android-native/app/src/main/java/in/juspay/mobility/utils/LocationUpdateService.java

```diff
@@ -368,8 +371,154 @@ private void updateDriverStatus(Boolean status) {
+
+            for (int i = 0; i < waypointArray.length(); i++) {
+                JSONObject pt = waypointArray.getJSONObject(i);
+                LatLngPoints.add(new LatLng((Double) pt.get("lat"), (Double) pt.get("lon")));
+            }
+
+
+            int resultIndex = PolyUtil.locationIndexOnEdgeOrPath(new LatLng(latitude, longitude), LatLngPoints, PolyUtil.isClosedPolygon(LatLngPoints), true, 30.0);
+
+            if (resultIndex == -1) {
+                int deviation = Integer.parseInt(sharedPreff.getString("RIDE_WAYPOINT_DEVIATION_COUNT", null));
+                deviation = deviation + 1;
+                SharedPreferences.Editor editor = sharedPreff.edit();
+                editor.putString("RIDE_WAYPOINT_DEVIATION_COUNT", String.valueOf(deviation));
+                editor.apply();
+
+                Log.d("new deviation count" , String.valueOf(deviation));
+            } else {
+               Log.d(  "in path", String.valueOf(resultIndex));
+            }
+        }
+        catch (Exception e) {
+            Log.e("deviation update exception", e.toString());
+        }
+    }
+
     /*Location update API call*/
-    private void callDriverCurrentLocationAPI(double latitude, double longitude, float accuracy, String locTime, String log, String locationSource, String triggerFunctionValue) {
+    private void callDriverCurrentLocationAPI(double latitude, double longitude, float accuracy, String locTime, String log, String locationSource, String triggerFunctionValue) throws JSONException {
+
+        try {
+            SharedPreferences sharedPreff = this.getSharedPreferences(this.getString(R.string.preference_file_key), Context.MODE_PRIVATE);
+            String localStage = sharedPreff.getString("LOCAL_STAGE", null);
+
+            if (localStage.equals("RideStarted")) {
+                String waypoints = sharedPreff.getString("RIDE_WAYPOINT", null);
+
+                if (waypoints.isEmpty()) {
+                    ExecutorService executor1 = Executors.newSingleThreadExecutor();
+                    executor1.execute(() -> {
+                        try {
+                            Handler routeHandler = new Handler(Looper.getMainLooper());
+                            String rideStartLat = sharedPreff.getString("RIDE_START_LAT", null);
+                            String rideStartLon = sharedPreff.getString("RIDE_START_LON", null);
+                            String rideEndLat = sharedPreff.getString("RIDE_END_LAT", null);
+                            String rideEndLon = sharedPreff.getString("RIDE_END_LON", null);
+
+                            System.out.println(Double.parseDouble(rideStartLat) + " " + Double.parseDouble(rideStartLon) + " " + Double.parseDouble(rideEndLat) + " " + Double.parseDouble(rideEndLon) + "fetched vals");
+
+                            getRoute(Double.parseDouble(rideStartLat), Double.parseDouble(rideStartLon), Double.parseDouble(rideEndLat), Double.parseDouble(rideEndLon));
```

Comment:
add a null check for the above values before making the API call. Else API call will be failed, some on other places wherever variable is being fetched from storage